File: Problems/NextPermutation.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var nextPermutation = function(nums) {
    let i, j;
    for (i = nums.length - 2; nums[i] >= nums[i + 1] && i >= 0; i--);
    if (i >= 0) {
        for (j = nums.length - 1; nums[j] <= nums[i]; j--);
            [nums[i], nums[j]] = [nums[j], nums[i]]; //swap
    }
    for (i++, j = nums.length - 1; i < j; i++, j--) {
            [nums[i], nums[j]] = [nums[j], nums[i]];
    }
};

File: Problems/NextPermutation.py
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        
        j = -1
        for i in range(len(nums) - 1):
            if nums[i] < nums[i+1]:
                j = i
        if j == -1:
            nums[:] = nums[::-1]
            return
        
        for i in range(j + 1, len(nums)):
            if nums[i] > nums[j]:
                k = i

        nums[j], nums[k] = nums[k], nums[j]
        
        nums[j+1:] = nums[:j:-1]
